File: python/src/feb/week2/boj/탑보기_박재환.py
def solution():
    import sys
    input = sys.stdin.readline

    n = int(input().strip())
    arr = [None for _ in range(n+1)]
    can_see = [0 for _ in range(n+1)]
    near = [0 for _ in range(n+1)]
    temp = list(map(int, input().split()))
    for i in range(1,n+1):
        arr[i] = (i, temp[i-1])

    def get_result():
        from collections import deque
        q = deque()

        for i in range(1, n+1):
            while q and q[-1][1] <= arr[i][1]:
                q.pop()
            can_see[i] += len(q)
            if q:
                near[i] = q[-1][0]
            q.append(arr[i])
        q.clear()
        for i in range(n, 0, -1):
            while q and q[-1][1] <= arr[i][1]:
                q.pop()
            can_see[i] += len(q)
            if q:
                candidate = q[-1][0]
                if near[i] == 0:
                    near[i] = candidate
                else:
                    dist_1 = abs(near[i] - arr[i][0])
                    dist_2 = abs(candidate - arr[i][0])

                    if dist_1 > dist_2:
                        near[i] = candidate
                    elif dist_1 == dist_2:
                        near[i] = min(near[i], candidate)
            q.append(arr[i])

        for i in range(1, n + 1):
            if can_see[i] == 0:
                print(0)
            else:
                print(can_see[i], near[i])

    get_result()
if __name__ == '__main__':
    solution()

File: java/src/feb/week2/boj/탑보기_박재환.java
package feb.week2.boj;

import java.util.*;
import java.io.*;

public class 탑보기_박재환 {
    /**
     * 다양한 높이의 건물 총 N개가 존재한다.
     * 건물높이를 L이라고 한다면, 높이가 L보다 큰 건물만 볼 수 있다.
     *
     * 바라보는 방향으로 높이가 L인 건물 뒤에 높이가 L이하인 건물이 있다면 가려져 보이지 않는다.
     */
    static BufferedReader br;
    static StringBuilder sb;
    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        sb = new StringBuilder();
        init();
        br.close();
        System.out.println(sb);
    }
    static StringTokenizer st;
    static int n;
    static Building[] arr;
    static void init() throws IOException {
        n = Integer.parseInt(br.readLine().trim());
        arr = new Building[n+1];
        st = new StringTokenizer(br.readLine().trim());
        for(int i=1; i<n+1; i++) arr[i] = new Building(i, Integer.parseInt(st.nextToken()));
        solution();
    }
    static class Building {
        int id, height;
        Building(int id, int height) {
            this.id = id;
            this.height = height;
        }
    }
    static void solution() {
        Deque<Building> stack = new ArrayDeque<>();
        int[] canSee = new int[n+1];
        int[] nearId = new int[n+1];

        // i번째 위치에서 왼쪽으로 볼 수 있는 빌딩
        for(int i=1; i<n+1; i++) {
            while(!stack.isEmpty() && stack.peek().height <= arr[i].height) stack.pop();
            canSee[i] += stack.size();
            if(!stack.isEmpty()) nearId[i] = stack.peek().id;
            stack.push(arr[i]);
        }
        stack.clear();
        // i번째 위치에서 오른쪽으로 볼 수 있는 빌딩
        for(int i=n; i>0; i--) {
            while(!stack.isEmpty() && stack.peek().height <= arr[i].height) stack.pop();
            canSee[i] += stack.size();
            if(!stack.isEmpty()) {
                int candidateLoc = stack.peek().id;
                if(nearId[i] == 0) nearId[i] = candidateLoc;
                else {
                    int dist1 = Math.abs(nearId[i] - arr[i].id);
                    int dist2 = Math.abs(candidateLoc - arr[i].id);

                    if (dist2 < dist1) nearId[i] = candidateLoc;
                    else if(dist1 == dist2) nearId[i] = Math.min(candidateLoc, nearId[i]);
                }
            }
            stack.push(arr[i]);
        }
        //--------------------------------------------------------------------------------------
        for(int i=1; i<n+1; i++) {
            if(canSee[i] == 0) sb.append(0);
            else {
                sb.append(canSee[i]).append(' ').append(nearId[i]);
            }
            sb.append('\n');
        }
    }
}
